File: src/engine.rs
use crate::book::OrderBook;
use crate::types::{Order, Side, Trade};

#[derive(Debug, Default)]
pub struct MatchingEngine {
    book: OrderBook,
}

impl MatchingEngine {
    pub fn new() -> Self {
        Self {
            book: OrderBook::new(),
        }
    }

    pub fn process(&mut self, mut incoming: Order) -> Vec<Trade> {
        let mut trades = Vec::new();
        let opposite_side = match incoming.side {
            Side::Buy => Side::Sell,
            Side::Sell => Side::Buy,
        };

        while incoming.remaining > 0 {
            let best_price = match opposite_side {
                Side::Buy => self.book.best_bid(),
                Side::Sell => self.book.best_ask(),
            };

            let price_crosses = match best_price {
                Some(price) => match incoming.side {
                    Side::Buy => incoming.price >= price,
                    Side::Sell => incoming.price <= price,
                },
                None => false,
            };

            if !price_crosses {
                break;
            }

            let resting_price = best_price.unwrap();
            let front = self.book.front_at(opposite_side, resting_price);
            if front.is_none() {
                break;
            }

            let resting = front.unwrap();
            let fill_qty = incoming.remaining.min(resting.remaining);
            incoming.remaining -= fill_qty;
            resting.remaining -= fill_qty;

            trades.push(Trade {
                buy_order_id: if incoming.side == Side::Buy {
                    incoming.id
                } else {
                    resting.id
                },
                sell_order_id: if incoming.side == Side::Sell {
                    incoming.id
                } else {
                    resting.id
                },
                price: resting.price,
                quantity: fill_qty,
            });

            if resting.remaining == 0 {
                self.book.pop_front_at(opposite_side, resting_price);
            }
        }

        if incoming.remaining > 0 {
            self.book.insert_resting(incoming);
        }

        trades
    }

    pub fn cancel(&mut self, order_id: u64) -> Option<Order> {
        self.book.remove(order_id)
    }

    pub fn book(&self) -> &OrderBook {
        &self.book
    }
}

File: cpp/src/engine.cpp
#include "engine.h"

std::vector<Trade> MatchingEngine::process(Order incoming) {
    std::vector<Trade> trades;
    Side opposite = (incoming.side == Side::Buy) ? Side::Sell : Side::Buy;

    while (incoming.remaining > 0) {
        auto best_price_opt = (opposite == Side::Buy) ? book.best_bid() : book.best_ask();
        if (!best_price_opt) break;
        Price best_price = *best_price_opt;
        bool price_crosses = false;
        if (incoming.side == Side::Buy) price_crosses = incoming.price >= best_price;
        else price_crosses = incoming.price <= best_price;
        if (!price_crosses) break;

        auto front_opt = book.front_at(opposite, best_price);
        if (!front_opt) break;
        Order* resting = *front_opt;
        Quantity fill_qty = std::min(incoming.remaining, resting->remaining);
        incoming.remaining -= fill_qty;
        resting->remaining -= fill_qty;

        Trade t;
        t.buy_order_id = (incoming.side == Side::Buy) ? incoming.id : resting->id;
        t.sell_order_id = (incoming.side == Side::Sell) ? incoming.id : resting->id;
        t.price = resting->price;
        t.quantity = fill_qty;
        trades.push_back(t);

        if (resting->remaining == 0) {
            book.pop_front_at(opposite, best_price);
        }
    }

    if (incoming.remaining > 0) {
        book.insert_resting(incoming);
    }

    return trades;
}

std::optional<Order> MatchingEngine::cancel(OrderId id) {
    return book.remove(id);
}
